README.md
# Images.weserv.nl

[![Author](https://img.shields.io/badge/author-andrieslouw-blue.svg)](https://github.com/andrieslouw)
[![Author](https://img.shields.io/badge/author-kleisauke-blue.svg)](https://github.com/kleisauke)
[![Source Code](https://img.shields.io/badge/source-andrieslouw/imagesweserv-blue.svg)](https://github.com/andrieslouw/imagesweserv)
[![Software License](https://img.shields.io/badge/license-BSD3-brightgreen.svg)](https://opensource.org/licenses/BSD-3-Clause)
[![Build Status](https://travis-ci.org/andrieslouw/imagesweserv.svg?branch=3.x)](https://travis-ci.org/andrieslouw/imagesweserv)
[![Test Coverage](https://codeclimate.com/github/andrieslouw/imagesweserv/badges/coverage.svg)](https://codeclimate.com/github/andrieslouw/imagesweserv/coverage)

Source code of images.weserv.nl, to be used on your own server(s). Images.weserv.nl leverages powerful libraries like [php-vips](https://github.com/jcupitt/php-vips) (for image handling and manipulation), [Guzzle](https://github.com/guzzle/guzzle) (for sending HTTP requests) and [URI](https://github.com/thephpleague/uri) (to create and manage URIs).

## Technologies used

- Linux, [nginx](https://github.com/nginx/nginx), [libvips](https://github.com/jcupitt/libvips) and [PHP](https://github.com/php/php-src) (without these technologies, this project would never have been possible)
- [Cloudflare](https://www.cloudflare.com/) (for caching and IP-blocking)
- [Memcached](https://github.com/memcached/memcached)/[Redis](https://github.com/antirez/redis) (for rate limiting)
- [OpenDNS](https://www.opendns.com/) (for DNS-filtering)

## Documentation

See our [wiki documentation](https://github.com/andrieslouw/imagesweserv/wiki) or [API reference](https://images.weserv.nl/) for information on using images.weserv.nl.

## Docker deployment

For information on Docker deployment, please read the [Docker installation instructions](DOCKER.md).

## Submitting Bugs and Suggestions

We track support tickets, issues and feature requests using the [GitHub issue tracker](https://github.com/andrieslouw/imagesweserv/issues).

## Credits
[![Andries Louw Wolthuizen](https://avatars2.githubusercontent.com/u/11487455?v=3&s=120)](https://github.com/andrieslouw) | [![Kleis Auke Wolthuizen](https://avatars2.githubusercontent.com/u/12746591?v=3&s=120)](https://github.com/kleisauke)
------------- | -------------
[Andries Louw Wolthuizen](https://github.com/andrieslouw) | [Kleis Auke Wolthuizen](https://github.com/kleisauke)

## License

The BSD 3-Clause License. Please see [LICENSE.md](https://github.com/andrieslouw/imagesweserv/blob/3.x/LICENSE.md) for more information.

## Privacy Policy

Please see [Privacy-Policy.md](https://github.com/andrieslouw/imagesweserv/blob/3.x/Privacy-Policy.md) for more information.

## Last but not least
This is made in Sneek with love and passion

[![Made in Sneek](https://kleisauke.nl/made-in-sneek-resized.png)](https://en.wikipedia.org/wiki/Sneek)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.0.0] - Unreleased

Requires libvips 8.7+ and PHP 7.1+.

### Note
With the feedback that you've provided we can successfully close the following issues/enhancements: 
[#10](https://github.com/andrieslouw/imagesweserv/issues/10), [#13](https://github.com/andrieslouw/imagesweserv/issues/13), [#15](https://github.com/andrieslouw/imagesweserv/issues/15), [#37](https://github.com/andrieslouw/imagesweserv/issues/37), [#62](https://github.com/andrieslouw/imagesweserv/issues/62), [#68](https://github.com/andrieslouw/imagesweserv/issues/68), [#69](https://github.com/andrieslouw/imagesweserv/issues/69), [#70](https://github.com/andrieslouw/imagesweserv/issues/70), [#75](https://github.com/andrieslouw/imagesweserv/issues/75), [#76](https://github.com/andrieslouw/imagesweserv/issues/76), [#78](https://github.com/andrieslouw/imagesweserv/issues/78), [#80](https://github.com/andrieslouw/imagesweserv/issues/80), [#81](https://github.com/andrieslouw/imagesweserv/issues/81), [#90](https://github.com/andrieslouw/imagesweserv/issues/90), [#106](https://github.com/andrieslouw/imagesweserv/issues/106), [#115](https://github.com/andrieslouw/imagesweserv/issues/115), [#121](https://github.com/andrieslouw/imagesweserv/issues/121), and [#122](https://github.com/andrieslouw/imagesweserv/issues/122). 
Thanks for your support!

### Added
#### Size
- Device pixel ratio `&dpr=`. See [#115](https://github.com/andrieslouw/imagesweserv/issues/115) for more info.

#### Transformation
- Letterboxing `&t=letterbox`. See [#80](https://github.com/andrieslouw/imagesweserv/issues/80) for more info.

#### Orientation
- Rotation `&or=`. Accepts `auto` or if an angle is specified, it is converted to a valid `90`/`180`/`270` degree rotation. Default is `auto`. The `auto` option uses Exif data to automatically orient images correctly.

#### Cropping
- Smart crop `&a=entropy` or `&a=attention` (only works when `&t=square`). Crops the image down to specific dimensions by removing boring parts. Where:
    - `entropy`: focus on the region with the highest [Shannon entropy](https://en.wikipedia.org/wiki/Entropy_%28information_theory%29).
    - `attention`: focus on the region with the highest luminance frequency, colour saturation and presence of skin tones.
- Focal point cropping `&a=crop-x%-y%` (only works when `&t=square`). Using two offset percentages, where `x%` is the horizontal offset and `y%` is the vertical offset.
- Shape cropping `&shape=`. Accepts:
    - `circle`
    - `ellipse`
    - `triangle`
    - `triangle-180`: Triangle tilted upside down
    - `pentagon`
    - `pentagon-180`: Pentagon tilted upside down
    - `hexagon`
    - `square`: Square tilted 45 degrees
    - `star`: 5-point star
    - `heart`

#### Adjustments
- The brightness of an image `&bri=`. Use values between `-100` and `+100`, where `0` represents no change.
- The contrast of an image `&con=`. Use values between `-100` and `+100`, where `0` represents no change.
- The gamma of an image `&gam=`. Use values between `1` and `3`. The default value is `2.2`, a suitable approximation for sRGB images.
- Sharpen an image `&sharp=`. Required format: `f,j,r`. Arguments:
    - Flat `f` - Sharpening to apply to flat areas. (Default: `1.0`)
    - Jagged `j` - Sharpening to apply to jagged areas. (Default: `2.0`)
    - Radius `r` - Sharpening mask to apply in pixels. (optional)
- The background color of an image `&bg=`. Can be used in combination with letterboxing. Accepts hexadecimal RGB and RBG alpha formats. See [#81](https://github.com/andrieslouw/imagesweserv/issues/81) for more info.

#### Effects
- The blur effect `&blur=`. Use values between `0` and `100`. See [#69](https://github.com/andrieslouw/imagesweserv/issues/69).
- The filter effect `&filt=`. Accepts `greyscale`, `sepia` or `negate`.

#### Input
- With the magical help of [libvips](https://github.com/jcupitt/libvips) and the PHP binding [php-vips](https://github.com/jcupitt/php-vips), we "officially" support `PNG`, `JPG`, `WEBP`, `GIF` (not animated), `SVG`, `PDF` and `TIFF` as image input. "Unofficially" we're supporting all [libMagick image file types](https://www.imagemagick.org/script/formats.php#supported). 

#### Output
- We've added `&output=webp` and `&output=tiff` in an effort to support more image formats as output. See [#68](https://github.com/andrieslouw/imagesweserv/issues/68).

#### Improvements
- Image filename in HTTP header (`Content-Disposition: inline`). See [#78](https://github.com/andrieslouw/imagesweserv/issues/78).
- The `&filename=` parameter to specify the filename returned in the `Content-Disposition` header. The filename must only contain alphanumeric characters. See [#122](https://github.com/andrieslouw/imagesweserv/issues/122).
- Support for Cyrillic and Arabic characters. See [#13](https://github.com/andrieslouw/imagesweserv/issues/13).
- The `&errorredirect=` parameter to redirect to a default image if the image URL is not found. The redirect URL must be formatted the same as the `?url=` parameter. See [#37](https://github.com/andrieslouw/imagesweserv/issues/37).
- In order to load a given page (for an PDF, TIFF and multi-size ICO file) we've added the `&page=` parameter. The value is numbered from zero.
- Support for non-standard ports. See [#10](https://github.com/andrieslouw/imagesweserv/issues/10).
- Added a privacy policy. See [Privacy-Policy.md](Privacy-Policy.md).
- Add support for Docker deployment. See the [Docker installation instructions](DOCKER.md).

### Changed

src/Manipulators/Shape.php
<?php

namespace AndriesLouw\imagesweserv\Manipulators;

use AndriesLouw\imagesweserv\Manipulators\Helpers\Utils;
use Jcupitt\Vips\Access;
use Jcupitt\Vips\Image;

/**
 * @property string $shape
 * @property string $circle
 * @property string $strim
 */
class Shape extends BaseManipulator
{
    /**
     * Perform shape image manipulation.
     *
     * @param Image $image The source image.
     *
     * @throws \Jcupitt\Vips\Exception
     *
     * @return Image The manipulated image.
     */
    public function run(Image $image): Image
    {
        $shape = $this->getShape();

        if ($shape !== null) {
            $width = $image->width;
            $height = $image->height;

            [$path, $xMin, $yMin, $maskWidth, $maskHeight] = $this->getSVGShape($width, $height, $shape);

            $preserveAspectRatio = $shape === 'ellipse' ? 'none' : 'xMidYMid meet';
            $svg = '<?xml version=\'1.0\' encoding=\'UTF-8\' standalone=\'no\'?>';
            $svg .= "<svg xmlns='http://www.w3.org/2000/svg' version='1.1' width='$width' height='$height'";
            $svg .= " viewBox='$xMin $yMin $maskWidth $maskHeight'";
            $svg .= " shape-rendering='geometricPrecision' preserveAspectRatio='$preserveAspectRatio'>";
            $svg .= $path;
            $svg .= '</svg>';

            $mask = Image::newFromBuffer($svg, '', [
                'access' => Access::SEQUENTIAL
            ]);

            $image = $this->cutout($mask, $image);

            // Crop the image to the mask dimensions;
            // if strim is defined and if it's not a ellipse
            if (isset($this->strim) && $shape !== 'ellipse') {
                [$left, $top, $trimWidth, $trimHeight] = $this->resolveShapeTrim(
                    $width,
                    $height,
                    $maskWidth,
                    $maskHeight
                );

                // If the trim dimensions is less than the image dimensions
                if ($trimWidth < $width || $trimHeight < $height) {
                    $image = $image->extract_area($left, $top, $trimWidth, $trimHeight);
                }
            }
        }

        return $image;
    }

    /**
     * Resolve shape
     *
     * @return string|null The resolved shape.
     */
    public function getShape(): ?string
    {
        if ($this->shape === 'circle' ||
            $this->shape === 'ellipse' ||
            $this->shape === 'hexagon' ||
            $this->shape === 'pentagon' ||
            $this->shape === 'pentagon-180' ||
            $this->shape === 'square' ||
            $this->shape === 'star' ||
            $this->shape === 'heart' ||
            $this->shape === 'triangle' ||
            $this->shape === 'triangle-180'
        ) {
            return $this->shape;
        }

        // Deprecated use shape=circle instead
        if (isset($this->circle)) {
            return 'circle';
        }

        return null;
    }

    /**
     * Get the SVG shape
     *
     * @param int $width
     * @param int $height
     * @param string $shape
     *
     * @return mixed[] [
     *      *SVG path*,
     *      *Left edge of mask*,
     *      *Top edge of mask*,
     *      *Mask width*,
     *      *Mask height*
     * ]
     *
     * @SuppressWarnings(PHPMD.NPathComplexity)
     */
    private function getSVGShape(int $width, int $height, string $shape): array
    {
        $min = min($width, $height);
        $outerRadius = $min / 2;
        $midX = $width / 2;
        $midY = $height / 2;

        if ($shape === 'ellipse') {
            // Ellipse
            return ["<ellipse cx='$midX' cy='$midY' rx='$midX' ry='$midY'/>", 0, 0, $width, $height];
        }

        if ($shape === 'circle') {
            // Circle
            $xMin = $midX - $outerRadius;
            $yMin = $midY - $outerRadius;
            return ["<circle r='$outerRadius' cx='$midX' cy='$midY'/>", $xMin, $yMin, $min, $min];
        }

        if ($shape === 'heart') {
            // Heart
            return $this->getSVGHeart($outerRadius, $outerRadius);
        }

        // 'inner' radius of the polygon/star
        $innerRadius = $outerRadius;

        // Initial angle (clockwise). By default, stars and polygons are 'pointing' up.
        $initialAngle = 0.0;

        // Number of points (or number of sides for polygons)
        $points = 0;

        switch ($shape) {
            case 'hexagon':
                // Hexagon
                $points = 6;
                break;
            case 'pentagon':
                // Pentagon
                $points = 5;
                break;
            case 'pentagon-180':
                // Pentagon tilted upside down
                $points = 5;
                $initialAngle = M_PI;
                break;
            case 'star':
                // 5 point star
                $points = 5 * 2;
                $innerRadius *= .382;
                break;
            case 'square':
                // Square tilted 45 degrees
                $points = 4;
                break;
            case 'triangle':
                // Triangle
                $points = 3;
                break;
            case 'triangle-180':
                // Triangle upside down
                $points = 3;
                $initialAngle = M_PI;
                break;
        }

        return $this->getSVGMask($midX, $midY, [$points, $outerRadius, $innerRadius, $initialAngle]);
    }

    /**
     * Formula from http://mathworld.wolfram.com/HeartCurve.html
     *
     * @param float $midX midX
     * @param float $midY midY
     *
     * @return mixed[] [
     *      *SVG path*,
     *      *Left edge of mask*,
     *      *Top edge of mask*,
     *      *Mask width*,
     *      *Mask height*
     * ]
     */
    public function getSVGHeart(float $midX, float $midY): array
    {
        $path = '';
        $xArr = [];
        $yArr = [];
        for ($t = -M_PI; $t <= M_PI; $t += 0.02) {
            $xPt = 16 * (sin($t) ** 3);
            $yPt = 13 * cos($t) - 5 * cos(2 * $t) - 2 * cos(3 * $t) - cos(4 * $t);

            $x = round($midX + $xPt * $midX);
            $y = round($midY - $yPt * $midY);
            $xArr[] = $x;
            $yArr[] = $y;
            $path .= "$x $y L";
        }
        $xMin = min($xArr);
        $yMin = min($yArr);

        return ["<path d='$path Z'/>", $xMin, $yMin, max($xArr) - $xMin, max($yArr) - $yMin];
    }

    /**
     * Inspired by this JSFiddle: https://jsfiddle.net/tohan/8vwjn4cx/
     * modified to support SVG paths
     *
     * @param float $midX midX
     * @param float $midY midY
     * @param array $parameters mask parameters
     *
     * @return mixed[] [
     *      *SVG path*,
     *      *Left edge of mask*,
     *      *Top edge of mask*,
     *      *Mask width*,
     *      *Mask height*
     * ]
     */
    private function getSVGMask(float $midX, float $midY, array $parameters): array
    {
        /**
         * @var int $points Number of points (or number of sides for polygons)
         * @var float $outerRadius 'outer' radius of the star
         * @var float $innerRadius 'inner' radius of the star (if equal to outerRadius, a polygon is drawn)
         * @var float $initialAngle Initial angle (clockwise). By default, stars and polygons are 'pointing' up.
         */
        [$points, $outerRadius, $innerRadius, $initialAngle] = $parameters;

        $path = '';
        $xArr = [];
        $yArr = [];
        for ($i = 0; $i <= $points; $i++) {
            $angle = $i * 2 * M_PI / $points - M_PI / 2 + $initialAngle;
            $radius = $i % 2 === 0 ? $outerRadius : $innerRadius;
            if ($i === 0) {
                $path = 'M';
                // If an odd number of points, add an additional point at the top of the polygon
                // -- this will shift the calculated center point of the shape so that the center point
                // of the polygon is at x,y (otherwise the center is mis-located)
                if ($points % 2 === 1) {
                    $path .= "0 $radius M";
                }
            } else {
                $path .= ' L';
            }
            $x = round($midX + $radius * cos($angle));
            $y = round($midY + $radius * sin($angle));
            $xArr[] = $x;
            $yArr[] = $y;
            $path .= "$x $y";
        }
        $xMin = min($xArr);
        $yMin = min($yArr);

        return ["<path d='$path Z'/>", $xMin, $yMin, max($xArr) - $xMin, max($yArr) - $yMin];
    }

    /**
     * Cutout src over dst
     *
     * @param Image $mask
     * @param Image $dst
     *
     * @throws \Jcupitt\Vips\Exception
     *
     * @return Image
     */
    public function cutout(Image $mask, Image $dst): Image
    {
        $maskHasAlpha = $mask->hasAlpha();
        $dstHasAlpha = $dst->hasAlpha();

        // we use the mask alpha if it has alpha
        if ($maskHasAlpha) {
            $mask = $mask->extract_band($mask->bands - 1, ['n' => 1]);
        }

        // split dst into an optional alpha
        $dstAlpha = $dst->extract_band($dst->bands - 1, ['n' => 1]);

        // we use the dst non-alpha
        if ($dstHasAlpha) {
            $dst = $dst->extract_band(0, ['n' => $dst->bands - 1]);
        }

        // the range of the mask and the image need to match .. one could be
        // 16-bit, one 8-bit
        $dstMax = Utils::maximumImageAlpha($dst->interpretation);
        $maskMax = Utils::maximumImageAlpha($mask->interpretation);

        if ($dstHasAlpha) {
            // combine the new mask and the existing alpha ... there are
            // many ways of doing this, mult is the simplest
            $mask = $mask->divide($maskMax)->multiply($dstAlpha->divide($dstMax))->multiply($dstMax);
        }
        // Not needed; after the thumbnail manipulator it's not an
        // 16-bit image anymore.
        /*elseif ($dstMax !== $maskMax) {
            // adjust the range of the mask to match the image
            $mask = $mask->divide($maskMax)->multiply($dstMax);
        }*/

        // append the mask to the image data ... the mask might be float now,
        // we must cast the format down to match the image data
        return $dst->bandjoin([$mask->cast($dst->format)]);
    }

    /**
     * Calculate the area to extract
     *
     * @param int $width
     * @param int $height
     * @param int $maskWidth
     * @param int $maskHeight
     *
     * @return mixed[]
     */
    public function resolveShapeTrim(int $width, int $height, int $maskWidth, int $maskHeight): array
    {
        $xScale = (float)($width / $maskWidth);
        $yScale = (float)($height / $maskHeight);
        $scale = min($xScale, $yScale);

        $trimWidth = $maskWidth * $scale;
        $trimHeight = $maskHeight * $scale;
        $left = (int)round(($width - $trimWidth) / 2);
        $top = (int)round(($height - $trimHeight) / 2);

        return [$left, $top, (int)round($trimWidth), (int)round($trimHeight)];
    }
}
